### PR TITLE
perf: request-path hot-path optimizations — Tier 3–4 (#149)

### LINKED ISSUE #156: perf: avoid holding the global accounts Mutex across the O(N) assessment (accounts.rs)
Tier 3 (concurrency / robustness) of #149.

`accounts.rs:143,191` holds a global `std::Mutex` across the O(N) assessment, so all providers contend on one map. Narrow the lock scope so the O(N) work runs without holding the global mutex.

Constraint: preserve correctness of multi-account failover.

Part of #149.

### LINKED ISSUE #158: perf: skip the per-account refresh lock when the token is still valid (adapters/responses/pool.rs)
Tier 3 (concurrency / robustness) of #149.

`adapters/responses/pool.rs:331` holds a per-account refresh lock even when the token is valid, serializing concurrent requests to one account. Fast-path valid tokens without taking the refresh lock.

Part of #149.

### LINKED ISSUE #159: perf: reduce per-selection allocations in select_order / is_fable_model (accounts.rs)
Tier 4 (minor) of #149.

`select_order` allocates 2N owned strings per selection, and `is_fable_model` lowercases per account (`accounts.rs`). Avoid the per-selection owned-string allocations and the repeated lowercasing.

Part of #149.

### LINKED ISSUE #160: perf: reuse per-frame timeout timers and keepalive boundary rotation (codex_ws, keepalive)
Tier 4 (minor) of #149.

Per-frame timeout timer registration in `codex_ws`/`keepalive`, plus `keepalive` per-byte boundary rotation. Reuse timers / avoid per-frame registration and per-byte rotation.

Part of #149.

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

Tier 3–4 of the request-path hot-path performance audit (#149). Four sub-issues, one commit each:

- **#156 (Tier 3)** — `AccountPool::select_order_inner` no longer holds the global `entries` Mutex across the O(N) quota assessment. The lock now covers only entry insertion, stale-quota expiry, and a minimal per-account state clone; `assess_quota` / `governing_weekly_reset` (pure computation) run **after** the lock is released. Multi-account failover ordering is unchanged.
- **#159 (Tier 4)** — `is_fable_model` is computed **once** per selection instead of ~2N times (removing per-account model lowercasing), threaded into the quota helpers as a precomputed `bool`. Provider-string construction is hoisted out of the per-account loop.
- **#158 (Tier 3)** — `resolve_or_cooldown` no longer takes the per-account refresh lock for still-valid tokens. `CodexAuthStore::get_valid_chatgpt` already single-flights refreshes at the auth layer (#157 / #168), so valid-token resolution now proceeds concurrently, while the expired-token refresh + atomic writeback stays serialized by the auth-layer path-keyed guard. The force-refresh (401 failover) path keeps its per-account lock (conservative scope).
- **#160 (Tier 4)** — `run_turn` reuses one pinned `tokio::time::Sleep` for the idle timeout instead of registering a fresh timer per frame; `BoundaryTracker::push` inspects at most the trailing 4 bytes of a chunk instead of rotating per byte.

## Notes

- The codex_ws `select!` polls the frame source **before** the idle timer (`biased`) to preserve the inner-future-first semantics of the `tokio::time::timeout` it replaces — a frame arriving as the deadline elapses is delivered, not dropped in favor of the timeout.

## Testing

- `cargo fmt --all --check` — clean
- `cargo clippy --all-targets --all-features -- -D warnings` — clean
- `cargo test --all-features --workspace` — green (645 lib tests + all integration suites)
- Regression tests added: valid-token resolution is not serialized by the pool refresh lock (#158); long-chunk boundary tracking ending in `\r\n\r\n` (#160).

## Docs

Internal performance refactors only — no change to observable behavior, config keys, endpoints, CLI, or provider/model semantics, so no `README.md` / `docs/` / `site/` update is required (per AGENTS.md).

Closes #156
Closes #158
Closes #159
Closes #160
Part of #149

---
## Summary by cubic
Speeds up the request path by reducing pool lock scope and per-account work, allowing concurrent credential resolution for valid ChatGPT tokens, and trimming Codex WS idle-timer overhead. No behavior changes — just faster hot paths.

- **Refactors**
  - Account selection: hold the `entries` mutex only for insertion/expiry/snapshot; run quota assessment and weekly reset after unlock. Compute `is_fable_model` once per selection; provider string hoisted. `QuotaState` is `Clone`; `assess_quota`/`governing_weekly_reset` take a precomputed `bool`.
  - Auth: `resolve_or_cooldown` skips the per-account refresh lock for valid tokens; expired-token refresh remains single-flighted in `CodexAuthStore::get_valid_chatgpt`. Tests ensure valid-token resolution doesn’t wait.
  - WS: `run_turn` reuses a pinned `tokio::time::Sleep` and uses biased `select` so frames at the deadline are delivered. `BoundaryTracker::push` bulk-copies only the trailing 4 bytes. Tests cover idle-timeout error surfacing and long-chunk/empty/3-byte boundary cases.

- **Bug Fixes**
  - Removed a flaky idle-timer reset test; kept the reliable timeout-after-silence coverage.

<sup>Written for commit 3c8cb570176fe9fc0c103f10d2da819d62ed2fff. Summary will update on new commits.</sup>

### CODE REVIEW

File: src/accounts.rs [hunk 1]
```diff
@@ -221,30 +222,35 @@ impl AccountPool {
...
+                    .as_ref()
+                    .map_or_else(QuotaAssessment::ignored, |quota| {
+                        assess_quota(quota, account, is_fable, pool, unix_now)
+                    });
+                let weekly_reset = quota
+                    .as_ref()
+                    .and_then(|quota| governing_weekly_reset(quota, is_fable));
+                (cooldown_until, assessment, weekly_reset)
+            })
+            .collect::<Vec<_>>();
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/accounts.rs:253]: [claude] Verified correct — the selection logic (`rotation`/`sticky`/`available_under` and the sort closures) only reads `snapshots[index]` for `index ∈ ident_reps`, so staging alias accounts does redundant lock/allocation work whose results are never read. Rather than churn this hot-path staging in a PR that is ready to merge, tracking it as a focused follow-up: #183. Thanks for the catch!

File: src/adapters/responses/codex_ws.rs [hunk 2]
```diff
@@ -1352,6 +1361,75 @@ mod tests {
...
+            "unexpected error: {}",
+            error.message
+        );
+        assert!(
+            events.recv().await.is_none(),
+            "the event channel ends after the transport error"
+        );
+
+        drop(release_server);
+        server.await.unwrap();
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/adapters/responses/codex_ws.rs:1434]: [claude] Good catch — added `turn_idle_timer_resets_after_each_frame` in c3279af: a paused-time test that delivers a frame one second before the original deadline, advances past that old deadline and asserts no timeout fired (the reset granted a fresh full window), then times out after renewed silence — pinning both the reset-after-activity semantics and the `biased` tie-breaking. Also hardened both paused-time tests so a fully-paused runtime cannot auto-advance to the idle timer before the first loopback frame arrives. Thanks!

[the remark this answers is not in the document]
Comment (amondnet) [reply src/adapters/responses/codex_ws.rs:1434]: [claude] Follow-up: the reset-after-activity test I added (`turn_idle_timer_resets_after_each_frame`) proved non-deterministic under CI load and was removed in 3c8cb57. Root cause is structural at this layer — tokio's paused runtime auto-advances to the next timer while both the test and `run_turn` park waiting on real mid-stream socket I/O, firing the idle timeout before the frame is delivered (passed locally with fast I/O, raced in CI). The reliable timeout-after-silence test stays. Proper reset coverage needs a mock frame source rather than a real socket — tracked as #184.
